File: apps/desktop/app/_state/dashboards.ts
import {
  DefaultError,
  UseQueryOptions,
  useMutation,
  useQuery,
  useQueryClient,
} from '@tanstack/react-query';
import { invoke } from '@tauri-apps/api/tauri';
import { useCallback } from 'react';
import { queryKeys, useCloseTab, useOpenTab, type Dashboard } from '.';

type State = { [key: string]: Dashboard };

export const useDashboards = <T = State>(
  options?: Omit<
    UseQueryOptions<State, DefaultError, T>,
    'queryKey' | 'queryFn'
  >,
) => {
  return useQuery({
    queryKey: queryKeys['dashboards'],
    queryFn: async () => {
      const dashboards = await invoke<Dashboard[]>(
        'plugin:ipc|list_dashboards',
      );
      return dashboards.reduce((acc, dashboard) => {
        acc[dashboard.id!.id.String] = dashboard;
        return acc;
      }, {} as State);
    },
    ...options,
  });
};

export const useDashboard = (id: string) => {
  return useDashboards({
    select: useCallback((dashboards: State) => dashboards[id], [id]),
  });
};

export const useUpsertDashboard = () => {
  const queryClient = useQueryClient();
  const openTab = useOpenTab();
  return useMutation({
    mutationKey: ['upsertDashboard'],
    mutationFn: async ({
      dashboard,
      focus = true,
    }: {
      dashboard: Dashboard;
      focus?: boolean;
    }) => {
      const _dashboard = await invoke<Dashboard>(
        'plugin:ipc|upsert_dashboard',
        { dashboard },
      );
      const id = _dashboard.id!.id.String;
      queryClient.setQueryData(queryKeys['dashboards'], (prevData: State) => {
        return {
          ...prevData,
          [id]: _dashboard,
        };
      });
      openTab(`dashboard://${id}`, focus);
      return _dashboard;
    },
  });
};

export const useDeleteDashboard = () => {
  const queryClient = useQueryClient();
  const closeTab = useCloseTab();
  return useMutation({
    mutationKey: ['deleteDashboard'],
    mutationFn: async ({ id }: { id: string }) => {
      await invoke('plugin:ipc|delete_dashboard', { id });
      closeTab(`dashboard://${id}`);
      queryClient.setQueryData(queryKeys['dashboards'], (prevData: State) => {
        delete prevData[id];
        return { ...prevData };
      });
    },
  });
};

File: apps/desktop/src-tauri/src/ipc/dashboards.rs
use super::prelude::*;

#[tauri::command]
pub async fn list_dashboards(state: State<'_, IpcState>) -> IpcResult<Vec<Dashboard>> {
  let dashboards = state.get_ng().await?.db.list_dashboard().await?;
  Ok(dashboards)
}

#[tauri::command]
pub async fn upsert_dashboard(
  dashboard: Dashboard,
  state: State<'_, IpcState>,
) -> IpcResult<Dashboard> {
  let dashboard = state
    .get_ng()
    .await?
    .db
    .upsert_dashboard(&dashboard)
    .await?;
  Ok(dashboard)
}

#[tauri::command]
pub async fn delete_dashboard(id: &str, state: State<'_, IpcState>) -> IpcResult<()> {
  state.get_ng().await?.db.delete_dashboard(id).await?;
  Ok(())
}
